content/match-types-spec.md

```diff
@@ -0,0 +1,579 @@
+    * If `q.Y` is abstract, fail as not specific.
+    * If `q.Y` is a class member:
+      * If `q` is a skolem type `∃α:X`, fail as not specific.
+      * Otherwise, compute `matchPattern(ti, q.Y, 0, scrutIsWidenedAbstract)`.
+    * Otherwise, the underlying type definition of `q.Y` is of the form `= U`:
+      * If `q` is a skolem type `∃α:X` and `U` refers to `α`, fail as not specific.
+      * Otherwise, compute `matchPattern(ti, U, 0, scrutIsWidenedAbstract)`.
+  * If `T` is a concrete type alias to a type lambda:
+    * Let `P'` be the beta-reduction of `P`.
+    * Compute `matchPattern(P', X, variance, scrutIsWidenedAbstract)`.
+
+#### Disjointness
+
+This proposal initially did not include a discussion of disjointness.
+After initial review, it became apparent that it should also provide a specification for the "provably disjoint" test involved in match type reduction.
+Additional study revealed that, while *specifiable*, the current algorithm is very ad hoc and severely lacks desirable properties such as preservation along subtyping (see towards the end of this section).
+
+Therefore, this proposal now also includes a proposed specification for "provably disjoint".
+To the best of our knowledge, it is strictly stronger than what is currently implemented, with one exception.
+
+The current implementation considers that `{ type A = Int }` is provably disjoint from `{ type A = Boolean }`.
+However, it is not able to prove disjointness between any of the following:
+
+* `{ type A = Int }` and `{ type A = Boolean; type B = String }` (adding another type member)
+* `{ type A = Int; type B = Int }` and `{ type B = String; type A = String }` (switching the order of type members)
+* `{ type A = Int }` and class `C` that defines a member `type A = String`.
+
+Therefore, we drop the very ad hoc case of one-to-one type member refinements.
+
+On to the specification.
+
+A scrutinee `X` is *provably disjoint* from a pattern `P` iff it is provably disjoint from the type `P'` obtained by replacing every type capture in `P` by a wildcard type argument with the same bounds.
+
+We note `X ⋔ Y` to say that `X` and `Y` are provably disjoint.
+Intuitively, that notion is based on the following properties of the Scala language:
+
+* Single inheritance of classes
+* Final classes cannot be extended
+* Sealed traits have a known set of direct children
+* Constant types with distinct values are nonintersecting
+* Singleton paths to distinct `enum` case values are nonintersecting
+
+However, a precise definition of provably-disjoint is complicated and requires some helpers.
+We start with the notion of "simple types", which are a minimal subset of Scala types that capture the concepts mentioned above.
+
+A "simple type" is one of:
+
+* `Nothing`
+* `AnyKind`
+* `p.C[...Xi]` a possibly parameterized class type, where `p` and `...Xi` are arbitrary types (not just simple types)
```

Comment:
I propose to treat `scala.compiletime.ops.int.S` as a parameterised class type for purposes of disjointness. Then will it will be considered disjoint to `0` and any other negative integer literal?

Reply:
I don't think we can pretend that it is a parameterized class type. It is not. For one thing, its existence would invalidate the `final` status of `scala.Int`, which then prevents `Int` from being disjoint from other traits, although currently it is.

If we want to add `S[n]` here, it needs to have its own category, and all the rules (and proof) need to be adapted to account for it.